File: ml-benchmark/autolifter/single-pass/max_dist_between_zeros.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;
Config SampleSize = 20;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

max_dist_between_zeros = fix (
  \f: Int -> List -> Int. \cs: Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} ->
    let cs = if == h 0 then 0 else + cs 1 in
    max cs (f cs t)
  end
) 0;

main = single_pass max_dist_between_zeros;

File: ml-benchmark/autolifter/single-pass/max_dist_between_zeros.ml
type lst = Cons of int * lst | Nil of unit

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) ->
      let r = run t in
      Cons (h, r)
  in
  fun xs -> v (run xs)

let inf = 100

let max a b = if a < b then b else a

let rec aux cs xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) ->
    let cs2 = if h = 0 then 0 else cs + 1 in
    let r = aux cs2 t in
    max cs2 r

let max_dist_between_zeros xs = aux 0 xs

let program = single_pass max_dist_between_zeros